### PR TITLE
Adds SchemaUtils to delta sharing spark client to check readCompatibility of schemas for streaming requests

### PR DESCRIPTION
**Pull request by linzhou-db**
Adds SchemaUtils to delta sharing spark client to check readCompatibility of schemas for streaming requests.

### CODE REVIEW

**Comment by linzhou-db** [comment]
@chakankardb Discussed with Ryan, we'll skip the schema check on server and let client do all the schema check.

File: spark/src/main/scala/io/delta/sharing/spark/util/SchemaUtils.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def toFieldMap(
+      fields: Seq[StructField],
+      caseSensitive: Boolean = false): Map[String, StructField] = {
+      val fieldMap = fields.map(field => field.name -> field).toMap
+      if (caseSensitive) {
+        fieldMap
+      } else {
+        CaseInsensitiveMap(fieldMap)
+      }
+    }
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/util/SchemaUtils.scala:47]: Do we use caseSensitive = true?
Should we avoid constructing a map first and then copying it into case insensitive form?

File: spark/src/main/scala/io/delta/sharing/spark/util/SchemaUtils.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            isDatatypeReadCompatible(e.keyType, n.keyType) &&
+            isDatatypeReadCompatible(e.valueType, n.valueType)
+        case (a, b) => a == b
+      }
+    }
+
+    def isStructReadCompatible: Boolean = {
+      val existingFieldMap = toFieldMap(existingSchema)
+      // scalastyle:off caselocale
+      val existingFieldNames = existingSchema.fieldNames.map(_.toLowerCase).toSet
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/util/SchemaUtils.scala:70]: caseInsensitiveMap uses toLowerCase(Locale.ROOT) which makes it locale insensitive. Should we follow that?

File: spark/src/test/scala/io/delta/sharing/spark/util/SchemaUtilsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      ("int", make(IntegerType)),
+      ("string", make(StringType)),
+      ("struct", make(new StructType().add("a", StringType))),
+      ("array", make(ArrayType(IntegerType))),
+      ("map", make(MapType(StringType, FloatType)))
+    )
+    test(s"change of datatype should fail read compatibility - $scenario") {
+      for (a <- schemas.keys; b <- schemas.keys if a != b) {
+        assert(!isReadCompatible(schemas(a), schemas(b)),
+          s"isReadCompatible should have failed for: ${schemas(a)}, ${schemas(b)}")
```
Comment (chakankardb) [inline spark/src/test/scala/io/delta/sharing/spark/util/SchemaUtilsSuite.scala:48]: Should we also test that if keys are equal then they are compatible?

File: spark/src/main/scala/io/delta/sharing/spark/util/SchemaUtils.scala (hunk 1 above)
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/util/SchemaUtils.scala:47]: We don't use caseSensitive = true, I removed it.

I think constructing from a map is the only available constructor of CaseInsensitiveMap.

File: spark/src/test/scala/io/delta/sharing/spark/util/SchemaUtilsSuite.scala (hunk 3 above)
Comment (linzhou-db) [reply spark/src/test/scala/io/delta/sharing/spark/util/SchemaUtilsSuite.scala:48]: added

**Review by chakankardb — APPROVED** [review]

**Pull request merged**